### DISCUSSION TITLE
xCDAT suddenly stops working

### DISCUSSION THREAD

**Discussion by gkb999** [xCDAT/xcdat#457] [2. Questions] [answered]
Hi, I was working with xcdat for the past week, and accidentally uninstalled the environment. 
I again re-installed a separate environment by deleting the old folders too, but I still get that the module isn't there/exist when I import.
What could be the reason?

**Comment by pochedls** [comment]
I'm not exactly sure why this would happen (or how to diagnose it). Maybe something got mixed up in the uninstall / reinstall. Can you install xcdat in a new environment (with a new name):

`conda create -n xcdat_new -c conda-forge xcdat`

**Comment by gkb999** [comment]
Thanks for getting back. I already installed in a new environmnet. Hoever, I again tried to install in xcdat_new. 
I used : conda create -n xcdat_new -c conda-forge xcdat xesmf : this creates a new environment and i tried to import xcdat and it says : NoModuleFound error

**Reply by pochedls** [reply]
I'm not totally sure why this would happen. Just to be sure – you did "activate" the environment (i.e., `conda activate xcdat_new`), right? Could you also provide the output of `conda list`?

**Reply by gkb999** [reply]
Oh yeah, I did activate, as I have been working with multiple environments and yeah, did the same way.

If this helps: please see the follwoing: 
******************
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: C:\Users\buradagi\Anaconda3\envs\xcdat_new

  added / updated specs:
    - xcdat


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    aws-c-auth-0.6.26          |       h9215439_2          90 KB  conda-forge
    aws-c-s3-0.2.8             |       h8113e7b_0          74 KB  conda-forge
    aws-c-sdkutils-0.1.9       |       h8a79959_0          50 KB  conda-forge
    aws-crt-cpp-0.19.9         |       ha4c2e88_1         223 KB  conda-forge
    aws-sdk-cpp-1.10.57        |       h75fde64_9         3.1 MB  conda-forge
    cftime-1.6.2               |  py311h59ca53f_1         176 KB  conda-forge
    cytoolz-0.12.0             |  py311ha68e1ae_1         330 KB  conda-forge
    lxml-4.9.2                 |  py311h5942461_0         1.0 MB  conda-forge
    lz4-4.3.2                  |  py311haddf500_0          75 KB  conda-forge
    msgpack-python-1.0.5       |  py311h005e61a_0          76 KB  conda-forge
    netcdf4-1.6.3              |nompi_py311h896b27b_102         360 KB  conda-forge
    pandas-2.0.1               |  py311hf63dbb6_0        12.8 MB  conda-forge
    pyarrow-11.0.0             |py311h6a6099b_14_cpu         2.9 MB  conda-forge
    setuptools-67.7.2          |     pyhd8ed1ab_0         569 KB  conda-forge
    vc-14.3                    |      h3d8a991_11          16 KB  conda-forge
    vs2015_runtime-14.34.31931 |      h4c5c07a_11         709 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        22.5 MB

The following NEW packages will be INSTALLED:

  arrow-cpp          conda-forge/win-64::arrow-cpp-11.0.0-h57928b3_14_cpu
  aws-c-auth         conda-forge/win-64::aws-c-auth-0.6.26-h9215439_2
  aws-c-cal          conda-forge/win-64::aws-c-cal-0.5.21-h7cda486_2
  aws-c-common       conda-forge/win-64::aws-c-common-0.8.14-hcfcfb64_0
  aws-c-compression  conda-forge/win-64::aws-c-compression-0.2.16-h8a79959_5
  aws-c-event-stream conda-forge/win-64::aws-c-event-stream-0.2.20-h5f78564_4
  aws-c-http         conda-forge/win-64::aws-c-http-0.7.6-h2545be9_0
  aws-c-io           conda-forge/win-64::aws-c-io-0.13.19-h0d2781e_3
  aws-c-mqtt         conda-forge/win-64::aws-c-mqtt-0.8.6-hd211e0c_12
  aws-c-s3           conda-forge/win-64::aws-c-s3-0.2.8-h8113e7b_0
  aws-c-sdkutils     conda-forge/win-64::aws-c-sdkutils-0.1.9-h8a79959_0
  aws-checksums      conda-forge/win-64::aws-checksums-0.1.14-h8a79959_5
  aws-crt-cpp        conda-forge/win-64::aws-crt-cpp-0.19.9-ha4c2e88_1
  aws-sdk-cpp        conda-forge/win-64::aws-sdk-cpp-1.10.57-h75fde64_9
  blosc              conda-forge/win-64::blosc-1.21.3-hdccc3a2_0
  bokeh              conda-forge/noarch::bokeh-3.1.0-pyhd8ed1ab_0
  brotlipy           conda-forge/win-64::brotlipy-0.7.0-py311ha68e1ae_1005
  bzip2              conda-forge/win-64::bzip2-1.0.8-h8ffe710_4
  c-ares             conda-forge/win-64::c-ares-1.18.1-h8ffe710_0
  ca-certificates    conda-forge/win-64::ca-certificates-2022.12.7-h5b45459_0
  certifi            conda-forge/noarch::certifi-2022.12.7-pyhd8ed1ab_0
  cf_xarray          conda-forge/noarch::cf_xarray-0.8.0-pyhd8ed1ab_0
  cffi               conda-forge/win-64::cffi-1.15.1-py311h7d9ee11_3
  cftime             conda-forge/win-64::cftime-1.6.2-py311h59ca53f_1
  click              conda-forge/noarch::click-8.1.3-win_pyhd8ed1ab_2
  cloudpickle        conda-forge/noarch::cloudpickle-2.2.1-pyhd8ed1ab_0
  colorama           conda-forge/noarch::colorama-0.4.6-pyhd8ed1ab_0
  contourpy          conda-forge/win-64::contourpy-1.0.7-py311h005e61a_0
  cryptography       conda-forge/win-64::cryptography-40.0.2-py311h28e9c30_0
  curl               conda-forge/win-64::curl-8.0.1-h68f0423_0
  cytoolz            conda-forge/win-64::cytoolz-0.12.0-py311ha68e1ae_1
  dask               conda-forge/noarch::dask-2023.4.0-pyhd8ed1ab_0
  dask-core          conda-forge/noarch::dask-core-2023.4.0-pyhd8ed1ab_0
  distributed        conda-forge/noarch::distributed-2023.4.0-pyhd8ed1ab_0
  freetype           conda-forge/win-64::freetype-2.12.1-h546665d_1
  fsspec             conda-forge/noarch::fsspec-2023.4.0-pyh1a96a4e_0
  gflags             conda-forge/win-64::gflags-2.2.2-ha925a31_1004
  glog               conda-forge/win-64::glog-0.6.0-h4797de2_0
  hdf4               conda-forge/win-64::hdf4-4.2.15-h1334946_6
  hdf5               conda-forge/win-64::hdf5-1.14.0-nompi_h918d9b7_103
  idna               conda-forge/noarch::idna-3.4-pyhd8ed1ab_0
  importlib-metadata conda-forge/noarch::importlib-metadata-6.6.0-pyha770c72_0
  importlib_metadata conda-forge/noarch::importlib_metadata-6.6.0-hd8ed1ab_0
  intel-openmp       conda-forge/win-64::intel-openmp-2023.1.0-h57928b3_46319
  jinja2             conda-forge/noarch::jinja2-3.1.2-pyhd8ed1ab_1
  krb5               conda-forge/win-64::krb5-1.20.1-heb0366b_0
  lcms2              conda-forge/win-64::lcms2-2.15-h3e3b177_1
  lerc               conda-forge/win-64::lerc-4.0.0-h63175ca_0
  libabseil          conda-forge/win-64::libabseil-20230125.0-cxx17_h63175ca_1
  libaec             conda-forge/win-64::libaec-1.0.6-h63175ca_1
  libarrow           conda-forge/win-64::libarrow-11.0.0-h8646f51_14_cpu
  libblas            conda-forge/win-64::libblas-3.9.0-16_win64_mkl
  libbrotlicommon    conda-forge/win-64::libbrotlicommon-1.0.9-hcfcfb64_8
  libbrotlidec       conda-forge/win-64::libbrotlidec-1.0.9-hcfcfb64_8
  libbrotlienc       conda-forge/win-64::libbrotlienc-1.0.9-hcfcfb64_8
  libcblas           conda-forge/win-64::libcblas-3.9.0-16_win64_mkl
  libcrc32c          conda-forge/win-64::libcrc32c-1.1.2-h0e60522_0
  libcurl            conda-forge/win-64::libcurl-8.0.1-h68f0423_0
  libdeflate         conda-forge/win-64::libdeflate-1.18-hcfcfb64_0
  libexpat           conda-forge/win-64::libexpat-2.5.0-h63175ca_1
  libffi             conda-forge/win-64::libffi-3.4.2-h8ffe710_5
  libgoogle-cloud    conda-forge/win-64::libgoogle-cloud-2.8.0-hf2ff781_1
  libgrpc            conda-forge/win-64::libgrpc-1.52.1-h32da247_1
  libhwloc           conda-forge/win-64::libhwloc-2.9.1-h51c2c0f_0
  libiconv           conda-forge/win-64::libiconv-1.17-h8ffe710_0
  libjpeg-turbo      conda-forge/win-64::libjpeg-turbo-2.1.5.1-hcfcfb64_0
  liblapack          conda-forge/win-64::liblapack-3.9.0-16_win64_mkl
  libnetcdf          conda-forge/win-64::libnetcdf-4.9.2-nompi_hc664c2b_104
  libpng             conda-forge/win-64::libpng-1.6.39-h19919ed_0
  libprotobuf        conda-forge/win-64::libprotobuf-3.21.12-h12be248_0
  libsqlite          conda-forge/win-64::libsqlite-3.40.0-hcfcfb64_1
  libssh2            conda-forge/win-64::libssh2-1.10.0-h9a1e1f7_3
  libthrift          conda-forge/win-64::libthrift-0.18.1-h9ce19ad_0
  libtiff            conda-forge/win-64::libtiff-4.5.0-h6c8260b_6
  libutf8proc        conda-forge/win-64::libutf8proc-2.8.0-h82a8f57_0
  libwebp-base       conda-forge/win-64::libwebp-base-1.3.0-hcfcfb64_0
  libxcb             conda-forge/win-64::libxcb-1.13-hcd874cb_1004
  libxml2            conda-forge/win-64::libxml2-2.10.4-hc3477c8_0
  libxslt            conda-forge/win-64::libxslt-1.1.37-h0192164_0
  libzip             conda-forge/win-64::libzip-1.9.2-h519de47_1
  libzlib            conda-forge/win-64::libzlib-1.2.13-hcfcfb64_4
  locket             conda-forge/noarch::locket-1.0.0-pyhd8ed1ab_0
  lxml               conda-forge/win-64::lxml-4.9.2-py311h5942461_0
  lz4                conda-forge/win-64::lz4-4.3.2-py311haddf500_0
  lz4-c              conda-forge/win-64::lz4-c-1.9.4-hcfcfb64_0
  m2w64-gcc-libgfor~ conda-forge/win-64::m2w64-gcc-libgfortran-5.3.0-6
  m2w64-gcc-libs     conda-forge/win-64::m2w64-gcc-libs-5.3.0-7
  m2w64-gcc-libs-co~ conda-forge/win-64::m2w64-gcc-libs-core-5.3.0-7
  m2w64-gmp          conda-forge/win-64::m2w64-gmp-6.1.0-2
  m2w64-libwinpthre~ conda-forge/win-64::m2w64-libwinpthread-git-5.0.0.4634.697f757-2
  markupsafe         conda-forge/win-64::markupsafe-2.1.2-py311ha68e1ae_0
  mkl                conda-forge/win-64::mkl-2022.1.0-h6a75c08_874
  msgpack-python     conda-forge/win-64::msgpack-python-1.0.5-py311h005e61a_0
  msys2-conda-epoch  conda-forge/win-64::msys2-conda-epoch-20160418-1
  netcdf4            conda-forge/win-64::netcdf4-1.6.3-nompi_py311h896b27b_102
  numpy              conda-forge/win-64::numpy-1.24.3-py311h0b4df5a_0
  openjpeg           conda-forge/win-64::openjpeg-2.5.0-ha2aaf27_2
  openssl            conda-forge/win-64::openssl-3.1.0-hcfcfb64_2
  orc                conda-forge/win-64::orc-1.8.3-hada7b9e_0
  packaging          conda-forge/noarch::packaging-23.1-pyhd8ed1ab_0
  pandas             conda-forge/win-64::pandas-2.0.1-py311hf63dbb6_0
  parquet-cpp        conda-forge/noarch::parquet-cpp-1.5.1-2
  partd              conda-forge/noarch::partd-1.4.0-pyhd8ed1ab_0
  pillow             conda-forge/win-64::pillow-9.5.0-py311hc67b2de_0
  pip                conda-forge/noarch::pip-23.1.1-pyhd8ed1ab_0
  psutil             conda-forge/win-64::psutil-5.9.5-py311ha68e1ae_0
  pthread-stubs      conda-forge/win-64::pthread-stubs-0.4-hcd874cb_1001
  pthreads-win32     conda-forge/win-64::pthreads-win32-2.9.1-hfa6e2cd_3
  pyarrow            conda-forge/win-64::pyarrow-11.0.0-py311h6a6099b_14_cpu
  pycparser          conda-forge/noarch::pycparser-2.21-pyhd8ed1ab_0
  pyopenssl          conda-forge/noarch::pyopenssl-23.1.1-pyhd8ed1ab_0
  pysocks            conda-forge/noarch::pysocks-1.7.1-pyh0701188_6
  python             conda-forge/win-64::python-3.11.3-h2628c8c_0_cpython
  python-dateutil    conda-forge/noarch::python-dateutil-2.8.2-pyhd8ed1ab_0
  python-tzdata      conda-forge/noarch::python-tzdata-2023.3-pyhd8ed1ab_0
  python_abi         conda-forge/win-64::python_abi-3.11-3_cp311
  pytz               conda-forge/noarch::pytz-2023.3-pyhd8ed1ab_0
  pyyaml             conda-forge/win-64::pyyaml-6.0-py311ha68e1ae_5
  re2                conda-forge/win-64::re2-2023.02.02-h63175ca_0
  setuptools         conda-forge/noarch::setuptools-67.7.2-pyhd8ed1ab_0
  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0
  snappy             conda-forge/win-64::snappy-1.1.10-hfb803bf_0
  sortedcontainers   conda-forge/noarch::sortedcontainers-2.4.0-pyhd8ed1ab_0
  tbb                conda-forge/win-64::tbb-2021.9.0-h91493d7_0
  tblib              conda-forge/noarch::tblib-1.7.0-pyhd8ed1ab_0
  tk                 conda-forge/win-64::tk-8.6.12-h8ffe710_0
  toolz              conda-forge/noarch::toolz-0.12.0-pyhd8ed1ab_0
  tornado            conda-forge/win-64::tornado-6.3-py311ha68e1ae_0
  tzdata             conda-forge/noarch::tzdata-2023c-h71feb2d_0
  ucrt               conda-forge/win-64::ucrt-10.0.22621.0-h57928b3_0
  urllib3            conda-forge/noarch::urllib3-1.26.15-pyhd8ed1ab_0
  vc                 conda-forge/win-64::vc-14.3-h3d8a991_11
  vs2015_runtime     conda-forge/win-64::vs2015_runtime-14.34.31931-h4c5c07a_11
  wheel              conda-forge/noarch::wheel-0.40.0-pyhd8ed1ab_0
  win_inet_pton      conda-forge/noarch::win_inet_pton-1.1.0-pyhd8ed1ab_6
  xarray             conda-forge/noarch::xarray-2023.4.2-pyhd8ed1ab_0
  xcdat              conda-forge/noarch::xcdat-0.5.0-pyhd8ed1ab_0
  xorg-libxau        conda-forge/win-64::xorg-libxau-1.0.9-hcd874cb_0
  xorg-libxdmcp      conda-forge/win-64::xorg-libxdmcp-1.1.3-hcd874cb_0
  xyzservices        conda-forge/noarch::xyzservices-2023.2.0-pyhd8ed1ab_0
  xz                 conda-forge/win-64::xz-5.2.6-h8d14728_0
  yaml               conda-forge/win-64::yaml-0.2.5-h8ffe710_2
  zict               conda-forge/noarch::zict-3.0.0-pyhd8ed1ab_0
  zipp               conda-forge/noarch::zipp-3.15.0-pyhd8ed1ab_0
  zlib               conda-forge/win-64::zlib-1.2.13-hcfcfb64_4
  zstd               conda-forge/win-64::zstd-1.5.2-h12be248_6


Proceed ([y]/n)? y


Downloading and Extracting Packages

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate xcdat_new
#
# To deactivate an active environment, use
#
#     $ conda deactivate

**Reply by pochedls** [reply]
Can you `import xarray`?

**Reply by gkb999** [reply]
Oh yeah, I'm processing my work on that now.

**Reply by pochedls** [reply]
I'm not totally sure what is going on. This thread suggests [checking permissions](https://stackoverflow.com/questions/14295680/unable-to-import-a-module-that-is-definitely-installed). I think it should be on disk at `C:\Users\buradagi\Anaconda3\envs\xcdat_new\lib/python3.11/site-packages/xcdat`.

@tomvothecoder / @jasonb5 - any ideas?

**Reply by gkb999** [reply]
If that was the case, the very first time I installed should have been the same.
Like, the error should have raised then as I didn't have to edit anywhere and could import directly and work with xcdat.
It is after I accidentally removed the environment at Anaconda navigator.
Anyway, I'm not sure why just xcdat doesn't work too. the installation says all are imported, I'm using the same environment to test too. Still, I can't import :|

**Reply by tomvothecoder** [reply]
Hi @GIRIJA-KALYANI, can you post the Python code you executed and its stack trace? 

> Thanks for getting back. I already installed in a new environmnet. Hoever, I again tried to install in xcdat_new. I used : conda create -n xcdat_new -c conda-forge xcdat xesmf : this creates a new environment and i tried to import xcdat and it says : NoModuleFound error

I noticed you're attempting install `xesmf` in a conda environment on a Windows platform. 

Are you trying to use the [horizontal regridding API](https://xcdat.readthedocs.io/en/stable/generated/xarray.Dataset.regridder.horizontal.html) with `xesmf` by any chance?  We mention in our [Getting Started/Installation Guide](https://xcdat.readthedocs.io/en/stable/getting-started.html#installation) that **`xesmf` is not supported on Windows**.

**Reply by gkb999** [reply]
I see, but I did install it with xesmf for the first time and it worked too.
I'm not using regridding for now - it all basic operations (climatology and stuff). I installed xcdat without xesmf too, but the error while importing module doesn't disappear.
Should I uninstall xcdat / remove the environment and try installing without xesmf?
Is there any work around to use xcdat in win platform now?

**Reply by tomvothecoder** [reply]
> Is there any work around to use xcdat in win platform now?

`xcdat` supports Windows platform. `xesmf` does not and if you try to use xcdat's horizontal regridding with `xesmf` you will get an error that `xesmf` isn't supported on Windows.

From what I understand, you are getting something like this:

```
>>> import xcdat
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'xcdat'
```

I would just create a fresh environment with `xcdat` only and try your import again.

```
conda create -n xcdat_env -c conda-forge xcdat
conda activate xcdat_env
```

**Reply by gkb999** [reply]
After following as suggested, created a new environment, xcdat_env, opened jupyter notebook, and tested.
Again the same error (refer screenshot)
![image](https://user-images.githubusercontent.com/7091088/234732591-f9c2b078-21ad-4735-9b56-1eaea5f703cc.png)

**Reply by pochedls** [reply]
Can you try to navigate to the directory above where xcdat is installed in your xcdat_test environment? I think it should be something like this:

`C:\Users\buradagi\Anaconda3\envs\xcdat_env\lib\python3.11\site-packages\`

From there, you should see an `xcdat` folder with files like `axis.py`, `bounds.py`, etc.

If you are in your xcdat environment and you are in the `site-packages` folder, you should be able to import xcdat.

**Reply by gkb999** [reply]
Hi, I navigated to "C:\Users\buradagi\Anaconda3\envs\xcdat_env\Lib\site-packages", and I can see:
![image](https://user-images.githubusercontent.com/7091088/234735817-e54534eb-ccb5-4f09-b096-0308816e728f.png)

**Reply by pochedls** [reply]
Can you do the following:

* Start iPython (or use a Jupyter terminal)
* `cd C:\Users\buradagi\Anaconda3\envs\xcdat_env\Lib\site-packages`
* `import xcdat`

If xcdat is imported, I think there is an issue with your path (or Anaconda adding to your path or your environment is not being activated however you are using it). 

One other question: are you using Jupyter? If so, did you re-create the kernel for your environment? 

Also...please respond to all parts of this comment.

**Reply by gkb999** [reply]
Hi, firstly, I started iPython and imported, its successful.
![image](https://user-images.githubusercontent.com/7091088/234799977-439d84ed-295c-4c07-aa4c-645e2c5bdc2c.png)
So, should be the issue within anaconda environment.
Yeah, I'm using jupyter and trying to import xcdat through jupyter notebook.
Now, how about " re-create the kernel for your environment"? This requires ? 

I think I responded to all sentences :)
Thanks for getting back and trying to sort this out by the way.

**Reply by pochedls** [reply]
Okay – are you able to `import xcdat` (after activating `xcdat_env`) even if you are not in the `site-packages` directory)? 

I suspect the issue is that your kernel (in Jupyter) does not have xcdat. There is limited guidance on setting up xcdat with Jupyter ([here](https://xcdat.readthedocs.io/en/stable/getting-started-hpc-jupyter.html)); if this doesn't work, is there someone at your institution that can help you configure your environment (I don't think the issue you are having is an xcdat problem)?

**Reply by gkb999** [reply]
Yes, I'm able to install even when I'm not in site-packages directory.

**Reply by tomvothecoder** [reply, marked as answer]
As @pochedls mentioned, this is likely an issue with your Jupyter instance not using the appropriate kernel (the conda env with xcdat).

**Reply by gkb999** [reply]
It is sorted with the creation on a new jupyter kernel and using it in notebook there.
Thanks for all the support team, it did help :)

**Reply by tomvothecoder** [reply]
Awesome! We are glad to help!  We encourage you to stay involved with xCDAT as you try it out.

I'll be closing this discussion as resolved.

**Reply by gkb999** [reply]
Yes, I just wanted to inform it worked, please close!!!
